File: token-validation-core/src/test/java/no/nav/security/token/support/core/IssuerMockWebServer.java
package no.nav.security.token.support.core;

import com.nimbusds.jose.util.IOUtils;
import okhttp3.*;
import okhttp3.mockwebserver.Dispatcher;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import okio.BufferedSink;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.URI;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Objects;

public class IssuerMockWebServer {
    private static final Logger log = LoggerFactory.getLogger(IssuerMockWebServer.class);
    private static final String DISCOVERY_PATH = "/.well-known/openid-configuration";
    private MockWebServer server;
    private MockWebServer proxyServer;
    private URL discoveryUrl;
    private URL proxyUrl;
    private final boolean startProxyServer;

    public IssuerMockWebServer() {
        this(true);
    }

    public IssuerMockWebServer(boolean startProxyServer) {
        this.startProxyServer = startProxyServer;
    }

    public void start() throws IOException {
        this.server = new MockWebServer();
        this.server.start();
        this.discoveryUrl = this.server.url(DISCOVERY_PATH).url();
        this.server.setDispatcher(new Dispatcher() {
            @Override
            public MockResponse dispatch(@NotNull RecordedRequest request) {
                log.debug("received request on url={} with headers={}", request.getRequestUrl(), request.getHeaders());
                log.debug("comparing path in request '{}' with '{}'", request.getRequestUrl().encodedPath(),
                    DISCOVERY_PATH);
                if (request.getRequestUrl().encodedPath().endsWith(DISCOVERY_PATH)) {
                    log.debug("returning well-known json data");
                    return wellKnownJson();
                } else {
                    log.error("path not found, returning 404");
                    return new MockResponse().setResponseCode(404);
                }
            }
        });

        this.proxyServer = new MockWebServer();
        this.proxyServer.setDispatcher(new ProxyDispatcher(HttpUrl.parse(discoveryUrl.toString())));
        if (startProxyServer) {
            this.proxyServer.start();
            this.proxyUrl = URI.create("http://localhost:" + this.proxyServer.getPort()).toURL();
        }
    }

    public void shutdown() throws IOException {
        server.shutdown();
        proxyServer.shutdown();
    }

    public URL getDiscoveryUrl() {
        return discoveryUrl;
    }

    public URL getProxyUrl() {
        return proxyUrl;
    }

    private static MockResponse mockResponse(String json) {
        return new MockResponse()
            .setResponseCode(200)
            .setHeader("Content-Type", "application/json;charset=UTF-8")
            .setBody(json);
    }

    private static MockResponse wellKnownJson() {
        try {
            String json = IOUtils.readInputStreamToString(
                IssuerMockWebServer.class.getResourceAsStream("/metadata.json"), StandardCharsets.UTF_8);
            return mockResponse(json);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public MockWebServer getServer() {
        return this.server;
    }

    public MockWebServer getProxyServer() {
        return this.proxyServer;
    }

    public boolean isStartProxyServer() {
        return this.startProxyServer;
    }

    public void setServer(MockWebServer server) {
        this.server = server;
    }

    public void setProxyServer(MockWebServer proxyServer) {
        this.proxyServer = proxyServer;
    }

    public void setDiscoveryUrl(URL discoveryUrl) {
        this.discoveryUrl = discoveryUrl;
    }

    public void setProxyUrl(URL proxyUrl) {
        this.proxyUrl = proxyUrl;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        IssuerMockWebServer that = (IssuerMockWebServer) o;
        return startProxyServer == that.startProxyServer &&
            Objects.equals(server, that.server) &&
            Objects.equals(proxyServer, that.proxyServer) &&
            Objects.equals(discoveryUrl, that.discoveryUrl) &&
            Objects.equals(proxyUrl, that.proxyUrl);
    }

    @Override
    public int hashCode() {
        return Objects.hash(server, proxyServer, discoveryUrl, proxyUrl, startProxyServer);
    }

    public String toString() {
        return "IssuerMockWebServer(server=" + this.getServer() + ", proxyServer=" + this.getProxyServer() + ", discoveryUrl=" + this.getDiscoveryUrl() + ", proxyUrl=" + this.getProxyUrl() + ", startProxyServer=" + this.isStartProxyServer() + ")";
    }

    static class ProxyDispatcher extends Dispatcher {
        private final OkHttpClient client;
        private final HttpUrl serverUrl;

        ProxyDispatcher(HttpUrl url) {
            serverUrl = url;
            client = new OkHttpClient.Builder().build();
        }

        @Override
        public MockResponse dispatch(final RecordedRequest recordedRequest) {
            Request.Builder requestBuilder = new Request.Builder()
                .url(serverUrl)
                .headers(recordedRequest.getHeaders())
                .removeHeader("Host");

            if (recordedRequest.getBodySize() != 0) {
                requestBuilder.method(recordedRequest.getMethod(), new RequestBody() {
                    @Override
                    public MediaType contentType() {
                        return MediaType.parse(recordedRequest.getHeader("Content-Type"));
                    }

                    @Override
                    public void writeTo(BufferedSink sink) throws IOException {
                        recordedRequest.getBody().clone().readAll(sink);
                    }

                    @Override
                    public long contentLength() {
                        return recordedRequest.getBodySize();
                    }
                });
            }
            Request request = requestBuilder.build();
            log.debug("created request to destination: {}", request);
            try (Response response = client.newCall(request).execute()) {
                ResponseBody body = response.body();
                if (body != null) {
                    MockResponse mockResponse = new MockResponse();
                    mockResponse.headers(response.headers());
                    mockResponse.setBody(body.string());
                    mockResponse.setResponseCode(response.code());
                    return mockResponse;
                } else {
                    MockResponse mockResponse = new MockResponse();
                    mockResponse.status("proxy error, response body from destination was null");
                    mockResponse.setResponseCode(500);
                    return mockResponse;
                }
            } catch (IOException e) {
                log.error("got exception when proxying request.", e);
                MockResponse mockResponse = new MockResponse();
                mockResponse.status("proxy error: " + e.getMessage());
                mockResponse.setResponseCode(500);
                return mockResponse;
            }
        }
    }
}

File: token-validation-core/src/test/kotlin/no/nav/security/token/support/core/IssuerMockWebServer.kt
package no.nav.security.token.support.core

import com.nimbusds.jose.util.IOUtils
import mockwebserver3.*
import okhttp3.Headers
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import java.io.IOException
import java.net.URI
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.*
import okio.BufferedSink
import org.slf4j.LoggerFactory

class IssuerMockWebServer(val startProxyServer: Boolean = true) {

    private val log = LoggerFactory.getLogger(IssuerMockWebServer::class.java)

    private lateinit var server: MockWebServer
    private var proxyServer: MockWebServer? = null
    lateinit var discoveryUrl: URL
    var proxyUrl: URL? = null
    private set

    @Throws(IOException::class)
    fun start() {
        server = MockWebServer().apply {
            start()
            discoveryUrl = url(DISCOVERY_PATH).toUrl()
            dispatcher = object : Dispatcher() {
                override fun dispatch(request: RecordedRequest): MockResponse {
                    log.debug("received request on url={} with headers={}", request.url, request.headers)
                    log.debug("comparing path in request '{}' with '{}'", request.url.encodedPath, DISCOVERY_PATH)
                    return if (request.url.encodedPath.endsWith(DISCOVERY_PATH) == true) {
                        log.debug("returning well-known json data")
                        wellKnownJson()
                    } else {
                        log.error("path not found, returning 404")
                        MockResponse(code = 404)
                    }
                }
            }
        }

        if (startProxyServer) {
            proxyServer = MockWebServer().apply {
                dispatcher = ProxyDispatcher(discoveryUrl.toString().toHttpUrlOrNull()!!)
                start()
                proxyUrl = URI.create("http://localhost:$port").toURL()
            }
        }
    }

    @Throws(IOException::class)
    fun shutdown() {
      //  server.shutdown()
       // proxyServer?.shutdown()
    }

    fun getServer(): MockWebServer = server
    fun getProxyServer(): MockWebServer? = proxyServer
    fun isStartProxyServer(): Boolean = startProxyServer

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false

        other as IssuerMockWebServer

        if (startProxyServer != other.startProxyServer) return false
        if (server != other.server) return false
        if (proxyServer != other.proxyServer) return false
        if (discoveryUrl != other.discoveryUrl) return false
        if (proxyUrl != other.proxyUrl) return false

        return true
    }

    override fun hashCode(): Int {
        return Objects.hash(server, proxyServer, discoveryUrl, proxyUrl, startProxyServer)
    }

    override fun toString(): String {
        return "IssuerMockWebServer(server=$server, proxyServer=$proxyServer, discoveryUrl=$discoveryUrl, proxyUrl=$proxyUrl, startProxyServer=$startProxyServer)"
    }

    companion object {
        private const val DISCOVERY_PATH = "/.well-known/openid-configuration"

        private fun mockResponse(json: String): MockResponse {
            return MockResponse(200,  Headers.Companion.headersOf("Content-Type", "application/json;charset=UTF-8"),json)

        }

        private fun wellKnownJson(): MockResponse {
            return try {
                val json = IOUtils.readInputStreamToString(IssuerMockWebServer::class.java.getResourceAsStream("/metadata.json"), StandardCharsets.UTF_8)
                mockResponse(json)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
    }

    class ProxyDispatcher(private val serverUrl: HttpUrl) : Dispatcher() {
        private val client = okhttp3.OkHttpClient.Builder().build()
        private val log = LoggerFactory.getLogger(ProxyDispatcher::class.java)


        override fun dispatch(request: RecordedRequest): MockResponse {
            val requestBuilder = Request.Builder()
                    .url(serverUrl)
                    .headers(request.headers)
                    .removeHeader("Host")

            if (request.bodySize != 0L) {
                requestBuilder.method(request.method, object : RequestBody() {
                    override fun contentType(): MediaType? {

                    return request.headers.get("Content-Type")?.toMediaTypeOrNull()
                    }

                    @Throws(IOException::class)
                    override fun writeTo(sink: BufferedSink) {
                        //request.body.clone().readAll(sink)
                    }

                    override fun contentLength(): Long {
                        return request.bodySize
                    }
                })
            }
            val req = requestBuilder.build()
            log.debug("created request to destination: {}", req)
            return try {
                client.newCall(req).execute().use { response ->
                        response.body?.let { body ->
                        MockResponse(response.code,response.headers,response.body.toString())
                } ?: MockResponse(code = 500)
                }
            } catch (e: IOException) {
                log.error("got exception when proxying request.", e)
                MockResponse(code = 500)
            }
        }
    }
}